metagraph/src/graph/representation/succinct/boss.cpp

```diff
@@ -245,50 +248,61 @@ void BOSS::serialize(std::ofstream &outstream) const {
     serialize_number_vector_raw(outstream, F_);
     serialize_number(outstream, k_);
     serialize_number(outstream, state);
-    outstream.flush();
 
     // write Wavelet Tree
     W_->serialize(outstream);
-    outstream.flush();
 
     // write last array
     last_->serialize(outstream);
     outstream.flush();
 }
 
-void BOSS::serialize(Chunk&& chunk, const std::string &filename) {
-    const std::string &fname = remove_suffix(filename, kExtension) + kExtension;
-
-    std::ofstream out(fname, std::ios::binary);
-    if (!out)
-        throw std::ofstream::failure("Error: Can't write to file " + fname);
+void BOSS::serialize(Chunk&& chunk, std::ofstream &out, State state) {
+    if (!out.good())
+        throw std::ofstream::failure("Error: Can't write to file");
 
     // write F values, k, and state
     serialize_number_vector_raw(out, chunk.F_);
     serialize_number(out, chunk.k_);
-    serialize_number(out, State::STAT);
-    out.flush();
+    serialize_number(out, state);
 
     // write Wavelet Tree
     // TODO: optimize
-    {
-        wavelet_tree_stat W(chunk.get_W_width(), chunk.W_);
-        chunk.W_.close(true); // remove the file
-        W.serialize(out);
-        out.flush();
-    }
+#define SERIALIZE_W(wt_type) { \
```

Comment:
why use a define instead of an inline templated helper?

Reply:
that would be another option, but both of them work, so it doesn't matter